### PR TITLE
Model parallel fix

### PR DESCRIPTION
**Pull request by kaixuanliu**
We find a series of model parallel issues recently. And we submit this PR to fix them. Pls help review, thx! @vasqu

### CODE REVIEW

**Comment by kaixuanliu** [comment]
Failed cases solved in this PR:
```
tests/models/bigbird_pegasus/test_modeling_bigbird_pegasus.py::BigBirdPegasusModelTest::test_model_parallel_beam_search
tests/models/pegasus_x/test_modeling_pegasus_x.py::PegasusXModelTest::test_model_parallel_beam_search
tests/models/trocr/test_modeling_trocr.py::TrOCRStandaloneDecoderModelTest::test_model_parallel_beam_search
tests/models/idefics/test_modeling_idefics.py::IdeficsModelIntegrationTest::test_inference_natural_language_visual_reasoning
tests/models/instructblip/test_modeling_instructblip.py::InstructBlipForConditionalGenerationDecoderOnlyTest::test_model_parallel_beam_search
tests/models/instructblipvideo/test_modeling_instructblipvideo.py::InstructBlipVideoForConditionalGenerationDecoderOnlyTest::test_model_parallel_beam_search
tests/models/blip_2/test_modeling_blip_2.py::Blip2ModelIntegrationTest::test_inference_opt_multi_accelerator
```

File: src/transformers/generation/utils.py [hunk 1]
```diff
@@ -2830,8 +2830,10 @@ def _sample(
                 next_tokens = torch.argmax(next_token_scores, dim=-1)
 
             # finished sentences should have their next token be a padding token
-            if has_eos_stopping_criteria:
-                next_tokens = next_tokens * unfinished_sequences + pad_token_id * (1 - unfinished_sequences)
+            if has_eos_stopping_criteria and pad_token_id is not None:
```
Comment (vasqu) [inline src/transformers/generation/utils.py]: This is a bit weird because it means we no longer pad finished sequences, imo it could be even some arbitrary token. Would be careful here cc @zucchini-nlp if you have some thoughts on this

**Review by vasqu — COMMENTED** [review]
I think most changes are fine, I'm only having some concerns in the generation utils to skip padding when we dont have a padding token id; this should be handled differently imo

File: src/transformers/generation/utils.py (hunk 1 above)
Comment (kaixuanliu) [reply src/transformers/generation/utils.py]: It is indeed somewhat weird here. Maybe we can put the `to` conversion of `pad_token_id` before the loop? This will have performance benefit as well.

Comment (vasqu) [reply src/transformers/generation/utils.py]: That's one part, but also `and pad_token_id is not None` condition is a bit weird - we should fill the sequences with some arbitrary token atp :thinking:

Comment (kaixuanliu) [reply src/transformers/generation/utils.py]: Usually it will be set to eos_token_id here: [pad_token_tensor = eos_token_tensor[0]](https://github.com/huggingface/transformers/blob/main/src/transformers/generation/utils.py#L1999C9-L2005C51), so actually it will never hit this condition, here I added this just to fix the LINT check by CI system

Comment (kaixuanliu) [reply src/transformers/generation/utils.py]: I submitted a new commit to refine the code here. Do you think it is OK?

File: src/transformers/generation/utils.py [hunk 2]
```diff
@@ -2763,6 +2763,13 @@ def _sample(
         this_peer_finished = False
         unfinished_sequences = torch.ones(batch_size, dtype=torch.long, device=input_ids.device)
 
+        # `pad_token_id` is created on `inputs_tensor.device` in `_prepare_special_tokens`. For multimodal models
+        # (e.g. BLIP-2, LLaVA) sharded across devices via `device_map="auto"`, `inputs_tensor` (e.g. `pixel_values`
+        # on the vision encoder) and `input_ids` (on the language model) can live on different devices, so we need to
+        # realign `pad_token_id` with `input_ids` to avoid cross-device ops below.
+        if pad_token_id is not None:
```
Comment (vasqu) [inline src/transformers/generation/utils.py:2770]: Just to double check but are we 100% sure that its a tensor at this point. (pretty sure it should but would be nice if you could confirm)

**Review by vasqu — APPROVED** [review]
Looks much more reasonable, thanks. Just one last question to confirm

File: src/transformers/generation/utils.py (hunk 2 above)
Comment (kaixuanliu) [reply src/transformers/generation/utils.py:2770]: Yes, it is torch.Tensor

**Comment by vasqu** [comment]
run-slow: bigbird_pegasus, idefics, instructblip, instructblipvideo, pegasus_x, trocr
